nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -99,33 +140,314 @@ fn skippable_version(
+
+            // This is the most "normal" case: Nexus should have access to the
+            // database, and the schema matches what it wants.
+            (HasExplicitAccess | HasImplicitAccess, UpToDate, _) => Self::Ready,
+
+            // If this Nexus is allowed to access the schema, but it looks
+            // older than what we expect, we'll need to update the schema to
+            // use it. Do this if and only if ConsumerPolicy allows it.
+            (
+                HasExplicitAccess | HasImplicitAccess,
+                OlderThanDesired | OlderThanDesiredSkipAccessCheck,
+                FailOnMismatch,
+            ) => Self::Refuse,
+            (
+                HasExplicitAccess | HasImplicitAccess,
+                OlderThanDesired | OlderThanDesiredSkipAccessCheck,
+                Update,
+            ) => Self::Update,
+        }
+    }
+}
+
 impl DataStore {
-    // Ensures that the database schema matches "desired_version".
-    //
-    // - Updating the schema makes the database incompatible with older
-    // versions of Nexus, which are not running "desired_version".
-    // - This is a one-way operation that cannot be undone.
-    // - The caller is responsible for ensuring that the new version is valid,
-    // and that all running Nexus instances can understand the new schema
-    // version.
-    //
-    // TODO: This function assumes that all concurrently executing Nexus
-    // instances on the rack are operating on the same version of software.
-    // If that assumption is broken, nothing would stop a "new deployment"
-    // from making a change that invalidates the queries used by an "old
-    // deployment".
-    pub async fn ensure_schema(
+    // Checks if the specified Nexus has access to the database.
+    async fn check_nexus_access(
         &self,
-        log: &Logger,
+        nexus_id: OmicronZoneUuid,
+    ) -> Result<NexusAccess, anyhow::Error> {
+        // Check if any "db_metadata_nexus" rows exist.
+        // If they don't exist, treat the database as having access.
+        //
+        // This handles the case for:
+        // - Fresh deployments where RSS hasn't populated the table yet (we need
+        // access to finish "rack_initialization").
+        // - Systems that haven't been migrated to include nexus access control
+        // (we need access to the database to backfill these records).
+        //
+        // After initialization/migration, this conditional should never trigger
+        // again.
+        let any_records_exist = self.database_nexus_access_any_exist().await?;
+        if !any_records_exist {
+            warn!(
+                &self.log,
+                "No db_metadata_nexus records exist - skipping access check";
+                "nexus_id" => ?nexus_id,
+                "explanation" => "This is expected during initial deployment or before migration"
+            );
+            return Ok(NexusAccess::HasImplicitAccess);
+        }
```

Comment:
I believe you've handled these cases separately via the RSS and schema update paths so I think this shouldn't be necessary.

Reply:
I believe that this check is actually necessary for the rack setup case to work correctly.

In the RSS case, we populate these records as a part of the "rack init" transaction -- but we need to get a connection to **perform** that transaction in the first place. If we can't bypass this check, we won't be able to call rack init.

The "migration to include nexus access control" case is a little fuzzier - the `schema-updater` is passing `IdentityCheckPolicy::DontCare`, so this will be skipped, and the schema update itself should be back-filling these records.